src/wallet/coin_selection.rs

```diff
@@ -138,6 +145,98 @@ impl CoinSelectionResult {
     }
 }
 
+/// Metric introduced to measure the performance of different coin selection algorithms.
+///
+/// This implementations considers waste as the sum of two values:
+/// * Timing cost
+/// * Creation cost
+/// > waste = timing_cost + creation_cost
+///
+/// **Timing cost** is the cost associated to the current fee rate and some long term fee rate used
+/// as a treshold to consolidate UTXOs.
+/// > timing_cost = txin_size * current_fee_rate - txin_size * long_term_fee_rate
+///
+/// Timing cost can be negative if the `current_fee_rate` is cheaper than the `long_term_fee_rate`,
+/// or zero if they are equal.
+///
+/// **Creation cost** is the cost associated to the surplus of coins besides the transaction amount
+/// and transaction fees. It can happen in the form of a change output or in the form of excess
+/// fees paid to the miner.
+///
+/// Change cost is derived from the cost of adding the extra output to the transaction and spending
+/// that output in the future.
+/// > change_cost = current_fee_rate * change_output_size + long_term_feerate * change_spend_size
+///
+/// Excess happens when there is not change, and the surplus of coins is spend as part of the fees
+/// to the miner:
+/// > excess = tx_total_value - tx_fees - target
+///
+/// Where _target_ is the desired amount to send.
+///
+/// Creation cost can be zero if there is a perfect match as result of the coin selection
+/// algorithm.
+///
+/// So, waste can be zero or negative if the creation cost is zero and the timing
+/// cost is less than or equal to zero.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Waste(pub i64);
+
+const LONG_TERM_FEE_RATE: FeeRate = FeeRate::from_sat_per_vb(10.0);
+
+impl Waste {
+    /// Calculate the amount of waste for the given coin selection
+    ///
+    /// - `selected`: the selected transaction inputs
+    /// - `cost_of_change`: the cost of creating change and spending it in the future
+    /// - `target`: the amount in satoshis of the aimed transaction output
```

Comment:
Possible documentation nit: I would expect the selection target to include the sum of the recipient amounts, the fees for creating the recipient outputs, and the fees for the (in practice) fixed transaction parts (version, locktime, input counter, output counter,  marker, flag). That way, only the inputs, the change, and additional fees from the inputs and `creation_cost` are up to be decided at the point of coin selection. It might be implemented differently here, but saying that `target` is just the recipient amounts might be slightly inaccurate.

Reply:
Not a nit at all, it's an important remark! I learn a few thing while reviewing
this, thanks!

It can be changed to:

> target: threshold in satoshis used to select UTXOs. It includes the sum of
> recipient outputs, the fees for creating the recipient outputs and the fees
> for fixed transaction parts

WDYT?

Reply:
SGTM